### ISSUE TITLE
Decouple Attack-Specific Logic from MINV User Input Handler

### ISSUE THREAD

**Issue by Hevaroo** [aidotse/LeakPro#248]
# Feature
We want to decouple attack-specific logic (such as PLGMI's GAN training) from user-defined input handlers (eg. celebA_plgmi_handler.py in examples). Currently, the input handler requires users to implement a training function specific to PLGMI, which makes it difficult to generalize to other attacks. Instead, the attack classes themselves should handle the data selection and any attack-specific components, while the user only provides a generic training function.

## Desired Behavior / Functionality

1. The user's input handler should not require knowledge of specific attacks, other than the opportunity to set attack parameters if desired.
2. The train_gan() function in the input handler should be completely attack-agnostic.

## What Needs to Be Done

1. Modify the train_gan() function in input handler: remove any assumptions about specific attack details. Perhaps also change name of this function.
2. Update prepare_attack in attack classes.

## How Can It Be Tested

1. Run existing attack(s) with this change. User input handler should not have any PLG-MI specific train logic.

**Comment by Hevaroo** [comment]
I feel like I need to discuss how we should set this up based on what makes sense for LeakPro. Please feel free to comment your ideas and thoughts @oscarwernqvist @johanos1 @fazelehh @henrikfo.

# How has it been done before?
In MIA, among some other things, you ask the user to also define a training loop. This makes a lot of sense here, since we for this attack may want to train some models that mimics how the target model behave on the incusion/exclusion of particular data points. Therefore, to ensure that these shadow models are a good approximation of this, they should be trained as similarly as possible as the target model. LeakPro cannot implement general code to ensure this. We cover this track by asking the user to input the training loop.

# Why does this not make sense for Model inversion reconstruction attacks?
Let's think about the intended use of the LeakPro framework. We have a user, that has a trained classification model (I think we only have considered classification models up to now...?), and is investigating how prone this model is to leaking data in a reconstruction setting. **_Then, why would the user have a generative model and/or training loop for this classification target model at hand?_**

Even if we assume that the user does have this, this would completely override the prepare_attack() function in the attacks, since it is in the choice of and training of the generative models that the attacks differ. This does however fall into a perfectly reasonable usage of the framework. The user might only be interested in the reconstruction part of the attack, namely run_attack() to get reconstructions and metrics.

# Suggestion
We make it so that it is not mandatory for the user to input a training loop, contrary to what has been done in MIA.
We provide the **option** for the user to provide a generative model **class** (GAN, Diffusion etc.).
We provide the **option** for the user to provide the **training loop** for the generative model provided.

We have some different scenarios:
1. The user inputs neither a generative model class nor a training loop. In this case, some default class would have to be used. The entirety of the desired attacks would have to run.
2. The user inputs a class, but no training loop. Here, the desired attacks would also run in their entirety. The provided generative class is here loaded in the attacks from user input, instead of some default.
3. The user inputs both a generative model class and a training loop. Here, prepare_attack is skipped in all of the desired attacks.

# Comments
I have some slight concern about scenario 1, because at first glance it might seem like that the burden of implementing default options to try to cover all use cases again falls on LeakPro (as in training loop for MIA). However this is perhaps not as bad, as the default generative model classes can more effectively cover the space of possible choices. Besides, there exists some excellent domain specific model architechtures, such as the CTGAN for tabular data, if one wished to run a GAN reconstruction attack for example.

This would be very straight-forward to implement with the new schemas structure!

# TLDR
Not requiring the user to input a (generative model) training loop (and associated generative model class) might make more sense in model inversion reconstruction attacks. This resolves the issue of abstracting attack specific logic from the user input handlers and minimizes requirements that is put on the user!

**Comment by oscarwernqvist** [comment]
I agree on that it's a lot of user input required for these attacks, however I still believe that we need to allow for all of these inputs in order to accomodate for all different types of attacks and types of data sets. Here are some thoughts:

1. If the user provides a pre-trained generator then we don't need to have a training loop for it, and in that case the user input is very straight-forward. In this case the user would still need to input the model class or at least reference to it with an import. Many of the attacks in the image domain use StyleGAN, so at least we can provide these attacks without a provided gan training loop.
2.  The examples provided will be important for the user, for more "textbook"-like applications such as images the existing GAN classes can be easily adapted, still with considerations to the image dimensions and augmentations.
3. For non "textbook" applications of model inversion attacks, I don't see how we can provide a GAN that is "plug-and-play".

**Comment by fazelehh** [comment]
I think LeakPro should support the first scenario where the user inputs neither a generative model class nor a training loop. As you already mentioned, this info is part of model inversion attacks and is not like training a target model.

**Comment by henrikfo** [comment]
I generally agree aswell with the sentiment on decoupling the attack-specific logic and above on 1. 

If LeakPro is provided with a generic (pre-trained) generator it should be using it when applicable. But when not provided it needs to have some default behavior.  

The pretrained case does not really make sense for example PLGMI. For similiar attack-procedures as PLGMI, you would like to train your generator with pseudo labels or atleast fine-tune it. This is attack specific, different attacks might have different, for example, losses and implementation details. 

LeakPro might have a generic training loop, given the attack-specific 
auxiliary data.

**Comment by henrikfo** [comment]
I generally agree aswell with the sentiment on decoupling the attack-specific logic and above on 1.

**Then, why would the user have a generative model and/or training loop for this classification target model at hand?**
Maybe we are talking about two different things at once here if I understand it. The training of a target model, or the already trained model should be given by the user, or atleast I think the assumption is that. We should of course give an example with a use-case and a training loop that trains a generic target model, but that is "outside" of any specific attack.

If LeakPro is provided with a generic (pre-trained) generator it should be using it when applicable. But when not provided OR the attack require something else, there needs to be some default behaviour. This should be provided by LeakPro. 

This default behaviour seems like it needs to be attack dependent. Different attacks have different "prepare_attack()" procedures, some use a pretrained GAN and some train it themselves. Then the fact that the pretrained generator does not really make sense for example in the case of PLGMI where you would like to train your generator with pseudo labels or atleast fine-tune it. This logic holds in general for similiar attack-procedures that differ from some sort of "base-case" if it even exists.  Further more, different attacks might have different losses and implementation details for its training. 

Of course you might want to re-use an already trained PLGMI-generator by loading from disk, just as its done for the target and shadow models in MIA. But thats another thing.

To go back and answer your question, a likely scenario would be that the user will not have a specified generator or training loop for the generator but will have thier own private model, or atleast have a training procedure for a model that is using their private data. I think the rule would be, and probably what makes sense is that prepare_attack() will issue an attack specific generator with an attack specific training loop _given_ a target model, or an ensemble of them. 

Said that, If there pops up more commonalities between most MINVA attacks, then it might make sense to extract that into separate functions some thing more generic. But we should also be mindful of technical debt and not try to do too much if not necessary.

This is just my two cents, feel free to disagree!
